pkg/controller/ec2/vpcendpoint/setup.go

```diff
@@ -0,0 +1,176 @@
+
+// SetupVPCEndpoint adds a controller that reconciles VPCEndpoint.
+func SetupVPCEndpoint(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(svcapitypes.VPCEndpointGroupKind)
+	opts := []option{
+		func(e *external) {
+			c := &custom{client: e.client, kube: e.kube}
+			e.delete = c.delete
+			e.preCreate = preCreate
+			e.postCreate = postCreate
+			e.postObserve = postObserve
+			e.isUpToDate = isUpToDate
+			e.filterList = filterList
+		},
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&svcapitypes.VPCEndpoint{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.VPCEndpointGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type custom struct {
+	kube   client.Client
+	client svcsdkapi.EC2API
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointInput) error {
+	// set external name as tag on the vpc endpoint
+	resType := "vpc-endpoint"
+	key := "Name"
+	value := meta.GetExternalName(cr)
+
+	spec := svcsdk.TagSpecification{
+		ResourceType: &resType,
+		Tags: []*svcsdk.Tag{
+			{
+				Key:   &key,
+				Value: &value,
+			},
+		},
+	}
+
+	obj.TagSpecifications = append(obj.TagSpecifications, &spec)
+	return nil
+}
+
+func postCreate(ctx context.Context, cr *svcapitypes.VPCEndpoint, obj *svcsdk.CreateVpcEndpointOutput, cre managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	// set vpc endpoint id as external name annotation on k8s object after creation
+	meta.SetExternalName(cr, aws.StringValue(obj.VpcEndpoint.VpcEndpointId))
+	cre.ExternalNameAssigned = true
+	return cre, nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.VPCEndpoint, resp *svcsdk.DescribeVpcEndpointsOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	// Load DNS Entry as connection detail
+	if len(resp.VpcEndpoints[0].DnsEntries) != 0 && awsclients.StringValue(resp.VpcEndpoints[0].DnsEntries[0].DnsName) != "" {
+		obs.ConnectionDetails = managed.ConnectionDetails{
+			xpv1.ResourceCredentialsSecretEndpointKey: []byte(awsclients.StringValue(resp.VpcEndpoints[0].DnsEntries[0].DnsName)),
+		}
+	}
+
+	cr.Status.AtProvider.VPCEndpoint = generateVPCEndpointSDK(resp.VpcEndpoints[0])
+
+	switch awsclients.StringValue(resp.VpcEndpoints[0].State) {
+	case "available":
+		cr.SetConditions(xpv1.Available())
+	case "pending":
+		cr.SetConditions(xpv1.Creating())
+	case "pending-acceptance":
```

Comment:
looks like these are available as constants, so you may consider using them to avoid typos: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2/types#VpcPeeringConnectionStateReasonCode

but I don't see `pending` there? should it be `VpcPeeringConnectionStateReasonCodeProvisioning`?

you can also use `case CodeFoo, CodeBar:` when two states should map to the same condition.

Reply:
It seems that those constants are for VPC Peering, and not for VPC Endpoints.

The closest constants that seems to match the state for VPC Endpoints are here: https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2/types#State

However, the casing in the constants are wrong. Hence, they were hard-coded instead. "pending" is indeed the correct state to be used for creating.

I'll fix the rest of the items.